src/NuGet.Clients/VsConsole/Console/ScriptExecutor.cs

```diff
@@ -0,0 +1,328 @@
+                }
+
+                string toolsPath = Path.GetDirectoryName(fullScriptPath);
+                IPSNuGetProjectContext psNuGetProjectContext = nuGetProjectContext as IPSNuGetProjectContext;
+                if (psNuGetProjectContext != null
+                    && psNuGetProjectContext.IsExecuting
+                    && psNuGetProjectContext.CurrentPSCmdlet != null)
+                {
+                    var psVariable = psNuGetProjectContext.CurrentPSCmdlet.SessionState.PSVariable;
+
+                    // set temp variables to pass to the script
+                    psVariable.Set("__rootPath", packageInstallPath);
+                    psVariable.Set("__toolsPath", toolsPath);
+                    psVariable.Set("__package", package);
+                    psVariable.Set("__project", envDTEProject);
+
+                    psNuGetProjectContext.ExecutePSScript(fullScriptPath, throwOnFailure);
+                }
+                else
+                {
+                    string logMessage = String.Format(CultureInfo.CurrentCulture, Resources.ExecutingScript, fullScriptPath);
+                    // logging to both the Output window and progress window.
+                    nuGetProjectContext.Log(MessageLevel.Info, logMessage);
+                    try
+                    {
+                        await ExecuteScriptCoreAsync(
+                            package,
+                            packageInstallPath,
+                            fullScriptPath,
+                            toolsPath,
+                            envDTEProject);
+                    }
+                    catch (Exception ex)
+                    {
+                        // throwFailure is set by Package Manager.
+                        if (throwOnFailure)
+                        {
+                            throw;
+                        }
+                        nuGetProjectContext.Log(MessageLevel.Warning, ex.Message);
+                    }
+                }
+
+                return true;
+            }
+            else
+            {
+                if (fullScriptPath.EndsWith(PowerShellScripts.Init, StringComparison.OrdinalIgnoreCase))
+                {
+                    TryMarkVisited(packageIdentity, false);
+                }
+            }
+            return false;
+        }
+
+        private async Task<string> GetPackageInstalledPathAsync(PackageIdentity packageIdentity)
+        {
+            // Since we need the solution directory when available, we switch to the main thread
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            string effectiveGlobalPackagesFolder = null;
+            if (!string.IsNullOrEmpty(SolutionDirectory))
+            {
+                var packagesFolder = PackagesFolderPathUtility.GetPackagesFolderPath(SolutionDirectory, Settings);
```

Comment:
resolving packages folder doesn't seem to belong here. There should only be one packages folder for this scenario, which is the global packages folder. This interface should be limited to that scenario and clearly document it